### ISSUE TITLE
Dart produces compile error whereas analyzer does not when trying to add `null` to `null`.

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#36343]
Dart version: 2.2.1-dev.1.1
OS: Windows 10

The following test example tries to calculate `null + null` expression:
``` dart
main() {
	try { null + null;} catch (e) {}
	print("OK");
}
```
Analyzer passes with this whereas dart unexpectedly produces compile time error.

Sample output is:

> $> dartanalyzer test.dart
> Analyzing test.dart...
> No issues found!
> $>
> $> dart test.dart
> test.dart:2:13: Error: The method '+' isn't defined for the class 'Null'.
> Try correcting the name to the name of an existing method, or defining a method named '+'.
> 	try { null + null;} catch (e) {}
> $>

Please note if I change the source for the following:
``` dart
main() {
	try { 1 + null;} catch (e) {}
	print("OK");
}
```
it works as expected: dart throws runtime exception, and so finally it prints "OK" as expected.

**Comment by iarkh** [comment]
One more failing example:
``` dart
main() {
	try {null >> 1;} catch (e) {}
	print("OK");
}
```

**Comment by lrhn** [comment]
The CFE is correct and the analyzer is wrong. The error is not entirely unexpected.

We changed the static type of `null` to be `Null` as part of the Dart 2 changes, rather than `bottom` as part of the preparation for Dart 2 ([CL](https://codereview.chromium.org/2555343004/diff/20001/docs/language/dartLangSpec.tex)). This was to prevent the `bottom` type from being propagated around by type inference (a `List<bottom>` is a curious beast). We also changed `Null` to be a proper sub-type of every other type. With non-nullable types, that too will cease to be the case.

A consequence of that is that `null + anything` is a static type error because `null` does not have any `+` method.
